### ISSUE TITLE
Improve course loading performance

### ISSUE THREAD

**Issue by dimkard** [kantord/LibreLingo#2314]
The load of large courses is a time-consuming process. My relatively old 4 x core workstation loads the basque course in ~ 34 seconds while on the pinephone the basque course loads after several minutes.

Using the cProfile profiling tool I get these results:

```
         113946968 function calls (113055904 primitive calls) in 64.162 seconds

   Ordered by: cumulative time
   List reduced from 840 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
      7/1    0.000    0.000   64.168   64.168 {built-in method builtins.exec}
        1    0.027    0.027   64.168   64.168 <string>:1(<module>)
        1    0.000    0.000   64.141   64.141 yaml_loader.py:469(load_course)
        1    0.000    0.000   63.857   63.857 yaml_loader.py:397(_load_modules)
        1    0.000    0.000   63.857   63.857 yaml_loader.py:401(<listcomp>)
       46    0.001    0.000   63.854    1.388 yaml_loader.py:363(_load_module)
       46    0.000    0.000   63.792    1.387 yaml_loader.py:351(_load_skills)
       46    0.011    0.000   63.792    1.387 yaml_loader.py:356(<listcomp>)
      144    0.005    0.000   63.773    0.443 yaml_loader.py:298(_load_skill)
      191    0.004    0.000   61.267    0.321 yaml_loader.py:38(_load_yaml)
      191    0.001    0.000   61.255    0.321 __init__.py:154(safe_load)
      191    0.184    0.001   61.254    0.321 __init__.py:103(load)
      191    0.001    0.000   61.049    0.320 constructor.py:47(get_single_data)
      191    0.001    0.000   57.717    0.302 composer.py:29(get_single_node)
      191    0.000    0.000   57.696    0.302 composer.py:50(compose_document)
654802/191    1.556    0.000   57.695    0.302 composer.py:63(compose_node)
198816/191    0.931    0.000   57.688    0.302 composer.py:117(compose_mapping_node)
16663/625    0.379    0.000   57.473    0.092 composer.py:99(compose_sequence_node)
  2387792    1.620    0.000   52.167    0.000 parser.py:94(check_event)
  5554398    4.056    0.000   39.989    0.000 scanner.py:113(check_token)
```

Regardless of the overhead that the profiler adds to the execution time, the root cause of the performance issues seems to be the `yaml.load` method.

**Possible solutions**

1. Port yaml files to json and use the json loader that I suppose that is more effective

2. Use an alternative yaml Loader.

If in the the [yaml_loader](https://github.com/LibreLingo/LibreLingo/blob/main/apps/librelingo_yaml_loader/librelingo_yaml_loader/yaml_loader.py) we change the yaml Loader to *CSafeLoader*, which is the SafeLoader alternative written in C using the LibYAML library., e.g.,:

```
def _load_yaml(path: Path):
    """Helper function for reading a YAML file"""
    with open(path) as yaml_file:
        return load(yaml_file, Loader=CSafeLoader) 
```
the execution time drops from ~34 seconds to ~6 seconds.

However, we need to ensure that CSafeLoader provides the same results as the SafeLoader and it is available in the platforms we use.

At the moment, I suppose a safe workaround can also be to use multiprocessing to load modules in parallel. In this case, the basque course loads in ~10 seconds.

#2313 contains this workaround.

**Comment by kantord** [comment]
>My relatively old 4 x core workstation loads the basque course in ~ 34 seconds while on the pinephone the basque course loads after several minutes.

even the 34 seconds is too long, but regarding the pinephone case, I would recommend you to generate static files instead of loading the course each time the app is loaded. Maybe pickling the loaded courses is an option?

**Comment by kantord** [comment]
> However, we need to ensure that CSafeLoader provides the same results as the SafeLoader and it is available in the platforms we use.

why would it not provide the same results? :scream_cat:  did you see some documentation that points to this? or just a general suspicion? 

Regarding it not being available, I suppose it's always possible to create a feedback version if it's not available, right?

**Comment by dimkard** [comment]
> why would it not provide the same results? scream_cat did you see some documentation that points to this? or just a general suspicion?

Well, investigating the issue I stumbled upon [this](https://github.com/emissary-ingress/emissary/issues/1294) comment. So, I gave it a try...

**Comment by dimkard** [comment]
> > My relatively old 4 x core workstation loads the basque course in ~ 34 seconds while on the pinephone the basque course loads after several minutes.
> 
> even the 34 seconds is too long, but regarding the pinephone case, I would recommend you to generate static files instead of loading the course each time the app is loaded. Maybe pickling the loaded courses is an option?

Yes, pickling the objects and creating a cache of loaded courses is a nice idea!

**Comment by kantord** [comment]
> > why would it not provide the same results? scream_cat did you see some documentation that points to this? or just a general suspicion?
> 
> Well, investigating the issue I stumbled upon [this](https://github.com/emissary-ingress/emissary/issues/1294) comment. So, I gave it a try...

actually, in your case even the safety is a bigger concern, right? I mean, for the web deployment process we can isolate the course build process from other parts of the deployment and completely isolate it from the app itself (no python involved there)

but your app actually loads the yaml files directly into the app, right?

**Comment by dimkard** [comment]
> but your app actually loads the yaml files directly into the app, right?

Exactly.

In specific, yamls and exports of courses available in Glossaico are [packaged](https://codeberg.org/dimkard/glossaico-data/tags) independently to decouple course content updates and Glossaico code updates.

When Glossaico starts, it checks [here](https://dimkard.codeberg.page/glossaico/configuration.yaml) for course content updates.

If updates are found, it fetches the updated yaml+exports archive and extracts it in the standard writable data directory of the system.

When the user selects a course, the yaml language course is loaded and pickled -as you suggested- and no further loads_course happens as long as there is no content update for this course.
